dsc_lib/src/functions/mod.rs

```diff
@@ -168,20 +188,43 @@ impl FunctionDispatcher {
-                return Err(DscError::Parser(t!("functions.noStringArgs", name = name, accepted_args_string = accepted_args_string).to_string()));
+
+        for (index, value) in args.iter().enumerate() {
+            if index >= metadata.accepted_arg_ordered_types.len() {
+                break;
+            }
+
+            if value.is_array() && !metadata.accepted_arg_ordered_types[index].contains(&FunctionArgKind::Array) {
+                return Err(DscError::Parser(t!("functions.noArrayArgs", name = name, accepted_args_string = metadata.accepted_arg_ordered_types[index].iter().map(std::string::ToString::to_string).collect::<Vec<_>>().join(", ")).to_string()));
+            } else if value.is_boolean() && !metadata.accepted_arg_ordered_types[index].contains(&FunctionArgKind::Boolean) {
+                return Err(DscError::Parser(t!("functions.noBooleanArgs", name = name, accepted_args_string = metadata.accepted_arg_ordered_types[index].iter().map(std::string::ToString::to_string).collect::<Vec<_>>().join(", ")).to_string()));
+            } else if value.is_null() && !metadata.accepted_arg_ordered_types[index].contains(&FunctionArgKind::Null) {
+                return Err(DscError::Parser(t!("functions.noNullArgs", name = name, accepted_args_string = metadata.accepted_arg_ordered_types[index].iter().map(std::string::ToString::to_string).collect::<Vec<_>>().join(", ")).to_string()));
+            } else if value.is_number() && !metadata.accepted_arg_ordered_types[index].contains(&FunctionArgKind::Number) {
+                return Err(DscError::Parser(t!("functions.noNumberArgs", name = name, accepted_args_string = metadata.accepted_arg_ordered_types[index].iter().map(std::string::ToString::to_string).collect::<Vec<_>>().join(", ")).to_string()));
+            } else if value.is_object() && !metadata.accepted_arg_ordered_types[index].contains(&FunctionArgKind::Object) {
+                return Err(DscError::Parser(t!("functions.noObjectArgs", name = name, accepted_args_string = metadata.accepted_arg_ordered_types[index].iter().map(std::string::ToString::to_string).collect::<Vec<_>>().join(", ")).to_string()));
+            } else if value.is_string() && !metadata.accepted_arg_ordered_types[index].contains(&FunctionArgKind::String) {
+                return Err(DscError::Parser(t!("functions.noStringArgs", name = name, accepted_args_string = metadata.accepted_arg_ordered_types[index].iter().map(std::string::ToString::to_string).collect::<Vec<_>>().join(", ")).to_string()));
+            }
+        }
+
+        // if we have remaining args, they must match one of the remaining_arg_types
+        if let Some(remaining_arg_types) = metadata.remaining_arg_accepted_types {
+            for value in args.iter().skip(metadata.accepted_arg_ordered_types.len()) {
+                if value.is_array() && !remaining_arg_types.contains(&FunctionArgKind::Array) {
```

Comment:
Might be worth encapsulating this logic in a helper function that checks if the value type is allowed based on a vec of arg types instead of duplicating for `metadata.accepted_arg_ordered_types[index]` & `remaining_arg_types`